controllers/cluster_controller.go

```diff
@@ -167,31 +166,34 @@ func (r *ClusterReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr e
 
 // reconcile handles cluster reconciliation.
 func (r *ClusterReconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster) (ctrl.Result, error) {
-	logger := r.Log.WithValues("cluster", cluster.Name, "namespace", cluster.Namespace)
-
-	// Call the inner reconciliation methods.
-	reconciliationErrors := []error{
-		r.reconcileInfrastructure(ctx, cluster),
-		r.reconcileControlPlane(ctx, cluster),
-		r.reconcileKubeconfig(ctx, cluster),
-		r.reconcileControlPlaneInitialized(ctx, cluster),
+	phases := []func(context.Context, *clusterv1.Cluster) (ctrl.Result, error){
+		r.reconcileInfrastructure,
+		r.reconcileControlPlane,
+		r.reconcileKubeconfig,
+		r.reconcileControlPlaneInitialized,
 	}
 
-	// Parse the errors, making sure we record if there is a RequeueAfterError.
 	res := ctrl.Result{}
 	errs := []error{}
-	for _, err := range reconciliationErrors {
-		if requeueErr, ok := errors.Cause(err).(capierrors.HasRequeueAfterError); ok {
-			// Only record and log the first RequeueAfterError.
-			if !res.Requeue {
-				res.Requeue = true
-				res.RequeueAfter = requeueErr.GetRequeueAfter()
-				logger.Error(err, "Reconciliation for Cluster asked to requeue")
-			}
-			continue
+	for _, fn := range phases {
+		// Call the inner reconciliation methods.
+		fnResult, err := fn(ctx, cluster)
+		if err != nil {
+			errs = append(errs, err)
```

Comment:
This can likely continue here, since errors take precedence over requeue:
```suggestion
			errs = append(errs, err)
			continue
```